### ISSUE TITLE
[Bug]: webgpu not support `shader-f16` feature

### ISSUE THREAD

**Issue by wszgrcy** [electron/electron#43567]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a bug report that matches the one I want to file, without success.

### Electron Version

32.0.1,30.1.2

### What operating system(s) are you using?

Windows

### Operating System Version

windows 10 19045.4717

### What arch are you using?

x64

### Expected Behavior

The 'shaderf16' feature of WebGPU is officially supported in [Chrome 120](https://developer.chrome.com/blog/new-in-webgpu-120)
I tested Chrome and chromium both support this feature
But when I tested the electron, I found that it doesn't have this feature. Even 32.0.1 doesn't have it
[test link](https://webgpureport.org/)
![image](https://github.com/user-attachments/assets/9f01d6e9-43db-436e-9193-0a91c4834f6a)

### Actual Behavior

![image](https://github.com/user-attachments/assets/4a20d9bb-f8fd-460b-951f-c187164e934a)

### Testcase Gist URL

https://gist.github.com/951547aa3c490a91ea30d1f773146fa4

### Additional Information

If this is a bug, can it be fixed to version 30 (Chrome 124)

**Comment by codebytere** [comment]
@wszgrcy I'm not sure I understand the issue - running this gist shows me the feature as expected:

<img width="912" alt="Screenshot 2024-09-11 at 3 24 10 PM" src="https://github.com/user-attachments/assets/618d570f-2d7a-4a33-b9b0-db6c58b95063">

**Comment by wszgrcy** [comment]
> @wszgrcy I'm not sure I understand the issue - running this gist shows me the feature as expected:
> 
> <img alt="Screenshot 2024-09-11 at 3 24 10 PM" width="912" src="https://private-user-images.githubusercontent.com/2036040/366474364-618d570f-2d7a-4a33-b9b0-db6c58b95063.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwNjMzOTgsIm5iZiI6MTcyNjA2MzA5OCwicGF0aCI6Ii8yMDM2MDQwLzM2NjQ3NDM2NC02MThkNTcwZi0yZDdhLTRhMzMtYjliMC1kYjZjNThiOTUwNjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MTFUMTM1ODE4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTAyNTc0N2M1OTU0ZDIxYzM3ZjA1OGIwZDQyODY4MDQzNWMxZDgyYzc2OTA0M2U0YjJmYWJhNTFhMTQ2M2VhZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.IZ53tyihAnQ_Lk3QqINshNDG7-K0rehKF_gekJGMqhw">

I'm not sure if it's related to the operating system. It's because I'm not displaying properly on Windows. And your screenshot seems to be from Mac?

**Comment by StevenHanbyWilliams** [comment]
This is only a windows issue.   Using the exact same project and dependency version on Mac and Windows produces different results, just for the shader-f16 feature.

chrome-version : 130.0.6723.59
node-version : 20.18.0
electron-version : 33.0.2

Windows 11  (note: Latest Drivers 566 )
![Image](https://github.com/user-attachments/assets/67ca1018-9a9f-4a97-96c8-0b15eb5f9c6e)

Mac OSX

![Image](https://github.com/user-attachments/assets/ec929597-ee93-43e8-b861-f81255fa2df3)

**Comment by wszgrcy** [comment]
bump

**Comment by deanihansen** [comment]
There's an open linux issue around Blink > WebGPU but nothing open to find on fp16 & Windows.
https://issues.chromium.org/issues/338730587

**Comment by ButzYung** [comment]
I can confirm that this problem still exists in electron 37 beta (Windows 10). FP16 support is important for some types of WebGPU applications, especially AI. FP16 model is half the size of a full FP32 model and faster in inference time.

**Comment by VerteDinde** [comment]
@deanihansen Sorry, getting caught up on the discussion here, but is this issue here that `shader-f16` isn't explicitly supported on Windows only in Electron, but is supported in Chromium? Looking at the original repro gist, is the expected behavior just that `shader-f16` should show up in the gist list?

Sorry, I'm not familiar with what `shader-f16` actually does; if someone here is able to provide more information about what you want/need to see outside of just the property appearing on the gist page itself, I can try to dig into why it's not appearing in Electron. Thanks!

**Comment by ButzYung** [comment]
@VerteDinde Yes, `shader-f16`, an additional feature commonly supported in most modern-day GPUs, will show up as supported in Chromium, but not in Windows Electron.

The following code will show what additional WebGPU features are supported.

`navigator.gpu.requestAdapter().then((adapter)=>{adapter.features.forEach((value)=>{ console.log(value) })})`

`shader-f16` will show up on the list in Chromium but not Windows Electron, just like the gist example above.

Of course it's more than just a missing part on the list. If you actually request the feature when creating the WebGPU device,

`navigator.gpu.requestAdapter().then((adapter)=>{adapter.requestDevice({requiredFeatures:['shader-f16']})})`

it will return the following error on Windows Electron.

> Uncaught (in promise) TypeError: Failed to execute 'requestDevice' on 'GPUAdapter': Unsupported feature: shader-f16

**Comment by deanihansen** [comment]
What @ButzYung said - what we'd love to see is electron webgpu supporting F16 shaders / half precision floating point / float16 / FP16. There's a bunch of abreviations and terms thrown around for the same thing. Chromium already supports it on Windows, so its definitely an Electron thing (maybe in Dawn or Blink if I recall correctly)..

Just found this chrome-dev post explaining it way better than could @VerteDinde 
[http://developer.chrome.com/blog/new-in-webgpu-120](http://developer.chrome.com/blog/new-in-webgpu-120)

**Comment by creeper-0910** [comment]
I've found an easy way to fix the issue! (This PR might not be accepted since I'm unsure about the DirectX Shader Compiler license, but at least it works...)

<img width="962" height="652" alt="Image" src="https://github.com/user-attachments/assets/d5de2be6-f8c3-4922-98b4-744eabd4d836" />

<img width="514" height="57" alt="Image" src="https://github.com/user-attachments/assets/149bb187-2f2d-43a6-baa2-7352b29128a7" />
